src/mono/System.Private.CoreLib/src/System/Reflection/TypeNameParser.Mono.cs

```diff
@@ -0,0 +1,165 @@
+        {
+            var name = new AssemblyName(assemblyName);
+
+            Assembly? assembly;
+            if (_assemblyResolver is not null)
+            {
+                assembly = _assemblyResolver(name);
+                if (assembly is null && _throwOnError)
+                {
+                    throw new FileNotFoundException(SR.Format(SR.FileNotFound_ResolveAssembly, assemblyName));
+                }
+            }
+            else
+            {
+                ref StackCrawlMark stackMark = ref Unsafe.AsRef<StackCrawlMark>(_stackMark);
+
+                if (_throwOnError)
+                {
+                    assembly = Assembly.Load(name, ref stackMark, null);
+                }
+                else
+                {
+                    // When throwOnError is false we should only catch FileNotFoundException.
+                    // Other exceptions like BadImangeFormatException should still fly.
+                    try
+                    {
+                        assembly = Assembly.Load(name, ref stackMark, null);
+                    }
+                    catch (FileNotFoundException)
+                    {
+                        return null;
+                    }
+                }
+            }
+            return assembly;
+        }
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2075:UnrecognizedReflectionPattern",
+            Justification = "TypeNameParser.GetType is marked as RequiresUnreferencedCode.")]
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "TypeNameParser.GetType is marked as RequiresUnreferencedCode.")]
+        private Type? GetType(string typeName, ReadOnlySpan<string> nestedTypeNames, string? assemblyNameIfAny)
+        {
+            Assembly? assembly = (assemblyNameIfAny is not null) ? ResolveAssembly(assemblyNameIfAny) : null;
+
+            // Both the external type resolver and the default type resolvers expect escaped type names
+            string escapedTypeName = EscapeTypeName(typeName);
+
+            Type? type;
+
+            // Resolve the top level type.
+            if (_typeResolver is not null)
+            {
+                type = _typeResolver(assembly, escapedTypeName, _ignoreCase);
+
+                if (type is null)
+                {
+                    if (_throwOnError)
+                    {
+                        throw new TypeLoadException(assembly is null ?
+                            SR.Format(SR.TypeLoad_ResolveType, escapedTypeName) :
+                            SR.Format(SR.TypeLoad_ResolveTypeFromAssembly, escapedTypeName, assembly.FullName));
+                    }
+                    return null;
+                }
+            }
+            else
+            {
+                if (assembly is null)
+                {
+                    ref StackCrawlMark stackMark = ref Unsafe.AsRef<StackCrawlMark>(_stackMark);
```

Comment:
`Unsafe.AsRef` is called on the value of `_stackMark`, not on its address. This can be re-written as: 
```
void* temp = _stackMark;
ref StackCrawlMark stackMark = ref Unsafe.AsRef<StackCrawlMark>(temp);
```
It should not be affected by the switch to `AutoLayout` in any way.